php/php.project/src/org/netbeans/modules/php/project/ComputeTestMethodAnnotations.java

```diff
@@ -0,0 +1,189 @@
+import org.netbeans.spi.project.SingleMethod;
+import org.openide.filesystems.FileObject;
+import org.openide.util.Lookup;
+import org.openide.util.RequestProcessor;
+
+
+public class ComputeTestMethodAnnotations implements DocumentListener, PropertyChangeListener, Runnable {
+
+    private static final Logger LOGGER = Logger.getLogger(ComputeTestMethodAnnotations.class.getName());
+
+    private static final RequestProcessor RP = new RequestProcessor(ComputeTestMethodAnnotations.class);
+    private final RequestProcessor.Task task = RP.create(this);
+
+    private static final ComputeTestMethodAnnotations INSTANCE = new ComputeTestMethodAnnotations();
+
+    private static final AtomicInteger usagesCount = new AtomicInteger(0);
+
+    private volatile Document handledDocument;
+
+    public static ComputeTestMethodAnnotations getInstance() {
+        return INSTANCE;
+    }
+
+    private ComputeTestMethodAnnotations() {
+    }
+
+    public void register() {
+        if (usagesCount.getAndIncrement() == 0) {
+            EditorRegistry.addPropertyChangeListener(this);
+        }
+    }
+
+    public void unregister() {
+        if (usagesCount.decrementAndGet() == 0) {
+            EditorRegistry.removePropertyChangeListener(this);
+        }
+    }
+
+    @Override
+    public void propertyChange(PropertyChangeEvent event) {
+        JTextComponent textComponent = EditorRegistry.lastFocusedComponent();
+        if (textComponent != null) {
+            Document document = textComponent.getDocument();
+            if (document != null) {
+                FileObject fileObject = NbEditorUtilities.getFileObject(document);
+                if (fileObject != null) {
+                    if (FileUtils.isPhpFile(fileObject)) {
+                        if (event.getPropertyName().equals(EditorRegistry.FOCUS_GAINED_PROPERTY)) {
+                            handleFileChange(document);
+                            document.addDocumentListener(ComputeTestMethodAnnotations.this);
+                        } else if (event.getPropertyName().equals(EditorRegistry.FOCUS_LOST_PROPERTY)) {
+                            document.removeDocumentListener(this);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void insertUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void removeUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void changedUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    private void handleFileChange(Document doc) {
+        handledDocument = doc;
+        task.schedule(500);
+    }
+
+    @Override
+    public void run() {
+        List<TestMethod> testMethods = new ArrayList<>();
+
+        List<TestMethod> methods = getTestMethods(handledDocument);
+        testMethods.addAll(methods);
```

Comment:
Why not directly `List<TestMethod> testMethods = getTestMethods(handledDocument);`?

Reply:
`TestMethodController.setTestMethods(handledDocument, getTestMethods(handledDocuemnt));`?

Reply:
Perhaps better as implemented - with a variable, we can check, whether it is empty (and avoid this call in such a case).